cardano-db-sync/src/Cardano/DbSync/Cache.hs

```diff
@@ -143,59 +132,60 @@ queryStakeAddrWithCache ::
   ReaderT SqlBackend m (Either (DB.LookupFail, ByteString) DB.StakeAddressId)
-queryStakeAddrWithCacheRetBs trce cache cacheUA nw cred = do
-  let !bs = Ledger.serialiseRewardAccount (Ledger.RewardAccount nw cred)
+queryStakeAddrWithCacheRetBs _trce cache cacheUA ra@(Ledger.RewardAccount _ cred) = do
+  let bs = Ledger.serialiseRewardAccount ra
   case cache of
     NoCache -> do
       mapLeft (,bs) <$> resolveStakeAddress bs
     ActiveCache ci -> do
-      prevCache <- liftIO $ readTVarIO (cStakeRawHashes ci)
-      let isNewCache = LRU.getSize prevCache < 1
-      -- populate from db if the cache is empty
-      currentCache <-
-        if isNewCache
-          then do
-            liftIO $ logInfo trce "Stake Raw Hashes cache is new and empty. Populating with addresses from db..."
-            queryRes <- DB.queryAddressWithReward (fromIntegral $ LRU.getCapacity prevCache)
-            liftIO $ atomically $ writeTVar (cStakeRawHashes ci) $ LRU.fromList queryRes prevCache
-            liftIO $ logInfo trce "Population of cache complete."
-            liftIO $ readTVarIO (cStakeRawHashes ci)
-          else pure prevCache
-
-      case LRU.lookup bs currentCache of
-        Just (addrId, lruCache) -> do
+      stakeCache <- liftIO $ readTVarIO (cStakeRawHashes ci)
+      case queryStakeCache cred stakeCache of
+        Just (addrId, stakeCache', _) -> do
           liftIO $ hitCreds (cStats ci)
           case cacheUA of
             EvictAndUpdateCache -> do
-              liftIO $ atomically $ writeTVar (cStakeRawHashes ci) $ LRU.delete bs lruCache
+              liftIO $ atomically $ writeTVar (cStakeRawHashes ci) $ deleteStakeCache cred stakeCache'
               pure $ Right addrId
             _other -> do
-              liftIO $ atomically $ writeTVar (cStakeRawHashes ci) lruCache
+              liftIO $ atomically $ writeTVar (cStakeRawHashes ci) stakeCache'
               pure $ Right addrId
         Nothing -> do
           queryRes <- mapLeft (,bs) <$> resolveStakeAddress bs
           liftIO $ missCreds (cStats ci)
           case queryRes of
             Left _ -> pure queryRes
             Right stakeAddrsId -> do
-              when (isCacheActionUpdate cacheUA) $
-                liftIO $
-                  atomically $
-                    modifyTVar (cStakeRawHashes ci) $
-                      LRU.insert bs stakeAddrsId
+              let !stakeCache' = case cacheUA of
+                    UpdateCache -> stakeCache {scLruCache = LRU.insert cred stakeAddrsId (scLruCache stakeCache)}
+                    UpdateCacheStrong -> stakeCache {scStableCache = Map.insert cred stakeAddrsId (scStableCache stakeCache)}
+                    _ -> stakeCache
+              liftIO $
+                atomically $
+                  writeTVar (cStakeRawHashes ci) stakeCache'
               pure $ Right stakeAddrsId
 
+-- | True if it was found in LRU
+queryStakeCache :: StakeCred -> StakeCache -> Maybe (DB.StakeAddressId, StakeCache, Bool)
```

Comment:
It doesn't look like the `Bool` is used at the call site.
Also how come we check `scStableCache` first isn't that more likely to be the bigger cache of the two?